KEP-5759: Memory Manager Hugepages Availability Verification

srikalyan opened the pull request:
## Summary

This KEP proposes enhancing the Memory Manager's Static policy to verify OS-reported free hugepages availability during pod admission.

## Problem

The Memory Manager only tracks hugepage allocations for Guaranteed QoS pods. Burstable/BestEffort pods can consume hugepages (via hugetlbfs mounts or `mmap` with `MAP_HUGETLB`) without being tracked, causing subsequent Guaranteed pods to be admitted but fail at runtime when hugepages are exhausted.

## Solution

1. **cadvisor**: Add `FreePages` field to `HugePagesInfo` with new `GetCurrentHugepagesInfo()` method for fresh sysfs reads (PR: https://github.com/google/cadvisor/pull/3804)
2. **Memory Manager**: Verify OS-reported free hugepages during `Allocate()` in Static policy
3. **Admission**: Reject pods when insufficient free hugepages are available

## Related

- **Enhancement Issue**: https://github.com/kubernetes/enhancements/issues/5759
- **Kubernetes Issue**: https://github.com/kubernetes/kubernetes/issues/134395
- **cadvisor PR**: https://github.com/google/cadvisor/pull/3804

## KEP Metadata

- **SIG**: sig-node
- **Stage**: Alpha (target v1.36)
- **Feature Gate**: `MemoryManagerHugepagesVerification`

/sig node
/kind kep

srikalyan committed:
KEP: Memory Manager Hugepages Availability Verification

This KEP proposes enhancing the Memory Manager's Static policy to
verify OS-reported free hugepages availability during pod admission.

Problem:
The Memory Manager only tracks hugepage allocations for Guaranteed QoS
pods. Burstable/BestEffort pods can consume hugepages without being
tracked, causing subsequent Guaranteed pods to be admitted but fail
at runtime when hugepages are exhausted.

Solution:
- Add FreePages field to cadvisor's HugePagesInfo (PR google/cadvisor#3804)
- Verify OS-reported free hugepages during Allocate() in Static policy
- Reject pods when insufficient free hugepages are available

Related: https://github.com/kubernetes/kubernetes/issues/134395

srikalyan commented:
/remove-area kubelet

ffromani commented:
/cc

srikalyan committed:
Fix TOC to pass verify-toc CI check

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+# KEP-NNNN: Memory Manager Hugepages Availability Verification
```
the KEP process is to file a issue on github.com/kubernetes/enhancements and use the issue number as KEP number. Example: https://github.com/kubernetes/enhancements/issues/5573

srikalyan replied:
Will do - I'll file an enhancement issue and update the KEP number accordingly.

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+This creates a problem when:
+1. Burstable or BestEffort pods consume hugepages (via hugetlbfs mounts or
+   `mmap` with `MAP_HUGETLB`) without being tracked by the Memory Manager
+2. External processes or other system components consume hugepages
```
One of the key assumptions of how the kubelet operates in general, is that is the sole owner of a node. There is some leeway in some cases: we can pre-partition CPUs and make the kubelet assume it is the sole owner of _the resource pool it got when started_, and we can probably should do the same for hugepages. But in general, dynamic co-sharing of resources (kubelet races with other daemons or programs) is not supported and it's unlikely it ever will

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+1. Burstable or BestEffort pods consume hugepages (via hugetlbfs mounts or
+   `mmap` with `MAP_HUGETLB`) without being tracked by the Memory Manager
+2. External processes or other system components consume hugepages
+3. The Memory Manager's internal state becomes stale or inconsistent with reality
```
how can this happen? do we have examples or scenarios?

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+### Goals
+
+- Verify OS-reported free hugepages during pod admission for the Static policy
+- Reject pods requesting hugepages when insufficient free hugepages are available
```
we need to be mindful this will cause another opportunity for rejection loops like https://github.com/kubernetes/kubernetes/issues/84869

ffromani replied:
we can't solve these cases without scheduler actions, and this would bring the total to at least 3 causes of runaway pod creation:
- TopologyAffinityError
- SMTAlignmentError
- now we are adding another error (HugePagesAlignmentError?).

Each of them makes sense from the node perspective - we can't admit a workload whose resource request can't be provided, but creates bad UX and storm of failed pods.

cc @44past4

srikalyan replied:
Agreed -- this is a valid concern. Admission rejection at the node level is the right thing to do (we can't admit what we can't provide), but without scheduler awareness of these node-level rejections, the same pod can be repeatedly scheduled to the same node, creating a storm of failures.

This is a broader kubelet/scheduler coordination problem that affects TopologyAffinityError and SMTAlignmentError equally. Solving it properly likely requires scheduler-side changes (e.g., back-off or taint-based feedback from repeated admission failures), which is beyond the scope of this KEP.

For this KEP, the admission error will follow the same pattern as existing node-level rejections, so it doesn't make the overall situation worse -- it just prevents a silent runtime failure from becoming the failure mode instead.

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+### Non-Goals
+
+- Track hugepage usage by Burstable or BestEffort pods in the Memory Manager
+- Modify scheduler behavior or add hugepage awareness to the scheduler
```
but we have already such awareness. The scheduler tracks hugepages already, so it becomes very important to clarify which scenarios lead to this desync

ffromani replied:
Makes sense from a strictly node perspective but we still have a gap at cluster level and we may need further work; or, at very least, clear docs (cc @dom4ha @44past4 FYI folks)

srikalyan replied:
Acknowledged. This KEP addresses the node-level gap where the kubelet/workload contract is breached. The cluster-level gap (scheduler not being aware of node-level hugepage exhaustion) is a broader issue shared with other admission errors like TopologyAffinityError.

I'll add a note to the Beta graduation criteria to include documentation on cluster-level behavior and known limitations, so operators understand what to expect when admission rejections occur.

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+- Track hugepage usage by Burstable or BestEffort pods in the Memory Manager
+- Modify scheduler behavior or add hugepage awareness to the scheduler
+- Provide hugepage reservation or preemption mechanisms
+- Support platforms other than Linux
```
We will likely need some conversation about this and sync with sig-node/sig-windows

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+As a database administrator, I run PostgreSQL with hugepages enabled for shared
+buffers. If an external monitoring agent or debugging tool temporarily consumes
+hugepages, subsequent Guaranteed pods requesting hugepages should not be admitted
+until hugepages are freed.
```
Arguably, a fix can be running these agents or tools as pod, so kubelet and kubernetes and the platform in general are aware of their presence and the accounting is correct. Even if we add the extra checks and more precise accounting, which I would generally in favor of, unexpected rejections are usually confusing the users.

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+
+- **sysfs dependency**: The feature depends on reading from sysfs. If sysfs is
+  unavailable or the free_hugepages file cannot be read, the feature gracefully
+  degrades to current behavior (no verification).
```
Reading and using free hugepages vs total hugepages as startup could be a good change, but I'd need to review how the validation currently works for hugepages. Can we add a summarization of what the current admission check in kubelet does?

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+| Risk | Mitigation |
+|------|------------|
+| sysfs reads add latency to admission | Minimal impact: single file read per hugepage size per NUMA node |
+| False rejections due to transient consumption | Acceptable: better to reject than admit and fail at runtime |
```
While I agree, it must be noted that this depends largely on the audience

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+|------|------------|
+| sysfs reads add latency to admission | Minimal impact: single file read per hugepage size per NUMA node |
+| False rejections due to transient consumption | Acceptable: better to reject than admit and fail at runtime |
+| sysfs unavailable in some environments | Graceful degradation: skip verification if sysfs unreadable |
```
On which supported Linux environments sysfs would not be available to the kubelet? pretty sure a bunch of other things would break anyway in this scenario

srikalyan replied:
Agreed - sysfs is always available on Linux systems where hugepages are configured. The graceful degradation language is perhaps overly defensive. I'll simplify the KEP to treat sysfs as reliably available on Linux.

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+    // number of huge pages
+    NumPages uint64 `json:"num_pages"`
+    // number of free huge pages (nil if unavailable)
+    FreePages *uint64 `json:"free_pages,omitempty"`
```
cadvisor maintainers will have the last word of course, but I'd not use a pointer here. Or, from the other angle: why do we need a pointer? I can't recall a platform or a kernel on which free hugepages reporting is not available.

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+#### Beta
+
+- E2e tests demonstrating correct behavior
+- Metrics for verification failures
```
I'd strongly prefer to have these at least partially available at alpha stage

srikalyan replied:
Agreed. I'll update the alpha graduation criteria to include at least partial e2e test coverage for the core scenario.

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+TBD during alpha phase.
```
I think we need an answer already for alpha stage

johnbelamaric replied:
can you fix this, then PRR is ready

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+  - Event Reason: `FailedHugepagesVerification`
+  - When: Pod admission rejected due to insufficient OS-reported free hugepages
+- [ ] Other
+  - Kubelet logs will indicate verification being performed and results
```
likewise, let's think about metrics or other explicit signals

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/kep.yaml:
```diff
@@ -0,0 +1,42 @@
+# The most recent milestone for which work toward delivery of this KEP has been
+# done. This can be the current (upcoming) milestone, if it is being actively
+# worked on.
+latest-milestone: "v1.33"
```
This should be 1.36, because we can't add features retroactively

srikalyan replied:
You're right. I'll update this to v1.36 since we can't add features retroactively. Thanks for catching this.

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/kep.yaml:
```diff
@@ -0,0 +1,42 @@
+
+# The milestone at which this feature was, or is targeted to be, at each stage.
+milestone:
+  alpha: "v1.33"
```
likewise, these will need to shifted using 1.36 as base

srikalyan replied:
Agreed. I'll shift all milestones using v1.36 as base:
- alpha: v1.36
- beta: v1.37  
- stable: v1.38

ffromani reviewed (commented):
Thanks for your contribution! I'm in favor of improving the accounting and making the memory manager/kubelet more predictable. I think we can benefit from some clarifications before to deep dive into further details.

srikalyan commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+###### How can an operator determine if the feature is in use by workloads?
+
+- Feature gate is enabled
+- Pods request hugepages resources
```
Agreed. I'll expand the metrics section and add specific proposals for metrics that would help operators monitor the feature's impact.

srikalyan committed:
Address reviewer feedback and close design gaps

Key changes:
- Update milestones to v1.36/v1.37/v1.38
- Clarify sysfs reading: add GetCurrentHugepagesInfo() for fresh reads
  (GetMachineInfo() is cached at startup, would be stale)
- Add Integration with Topology Manager section with policy behavior table
- Add Interaction with CPU Manager section
- Address reserved hugepages (free_hugepages is correct metric)
- Expand race condition discussion with failure handling details
- Rewrite Story 2 as "Rapid Pod Churn" with clear timeline
- Add "Static policy only" note (None policy not applicable)
- Specify error message format with example
- Add kubelet restart behavior note
- Update Risks table with new mitigations
- Fix unit test description (removed nil reference)
- Update TOC with new sections
- Link enhancement issue #5759

Related: https://github.com/kubernetes/enhancements/issues/5759

ffromani commented:
/retitle KEP-5759: Memory Manager Hugepages Availability Verification

ffromani commented on keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+
+As a cluster administrator running DPDK-based network functions, I deploy a
+Burstable pod that mounts hugetlbfs and consumes 2GB of 1GB hugepages for packet
+buffer pools. Later, I deploy a Guaranteed pod also requesting 2GB of 1GB hugepages.
```
ok, this makes sense. non-guaranteed pods don't have a NUMA affinity, so the problem is mapping those resources to NUMA nodes in order to fix the per-NUMA accounting, which is key in the guaranteed pods case. Mmeory accounting is however different from CPU time, and this is even truer for hugepages, which are a pretty static mapping. I'm wondering if tracking hugepages for all QoS pods does indeed  increase the scope much and if leads to a better fix. I think it's unlikely it will improve the rejection loops though.

srikalyan committed:
Present implementation options without recommendation

- Add two implementation approaches: Option A (direct sysfs) and Option B (cadvisor)
- Present pros/cons for each option neutrally for KEP review
- Remove cadvisor-specific sections, replace with options discussion
- Add Observability section with metrics, events, logs, alerting
- Update TOC to pass CI verification
- Update KEP number to 5759 throughout

The choice between implementation approaches is left to KEP reviewers
based on maintainability preferences and timeline considerations.

ffromani commented:
Thanks @srikalyan for leading this effort. I'm in general supportive of this memory manager enhancement and, pending further review and elaborating, I do see the benefit of the proposed approach about checking free hugepages. Because there's some time left before the 1.36 cycle begins, I'd like to explore other options to solve this problem before we commit to the proposed direction. I'll have another review iteration ASAP.

ffromani commented on keps/sig-node/5759-memory-manager-hugepages-verification/kep.yaml:
```diff
@@ -0,0 +1,43 @@
+participating-sigs: []
+status: provisional
+creation-date: 2024-12-24
+reviewers:
```
I can be a reviewer

ffromani commented on keps/sig-node/5759-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,478 @@
+
+**Desired behavior**: The Guaranteed pod admission fails immediately with a clear
+error indicating insufficient free hugepages, allowing the scheduler to try
+another node or the administrator to take corrective action.
```
the alternative I'm thinking about is to extend the memory manager and admission logic to listen to each and every pod admission to track where (= which NUMA node) the hugepages are allocated from. However, If the kubelet doesn't enforce a cpuset.mems restriction, however, there's no way to know from where the hugepage is gonna be taken till the container processes go running, therefore past admission stage. Therefore, the proposed approach to check the actual free resources before each and every allocation attempt seems to be the best compromise (bar the only possible approach) in the current architecture.
We should probably document this in the "discarded alternatives" section.

ffromani commented on keps/sig-node/5759-memory-manager-hugepages-verification/kep.yaml:
```diff
@@ -0,0 +1,43 @@
+
+# The milestone at which this feature was, or is targeted to be, at each stage.
+milestone:
+  alpha: "v1.36"
```
pending SIG discussion and approval, there's a good chance this work can start as beta per recent KEP graduation guidelines. The change is quite self contained and targeted, so it qualifies.

ffromani replied:
We got informal agreement in sig-node to start in beta

srikalyan replied:
Yes, targeting alpha for v1.36. The milestones have been relaxed to beta v1.38 and stable v1.40 per your feedback on the timeline being too aggressive.

ffromani reviewed (commented):
thanks for the updates. The next step is to bring this up on the larger sig-node and in the 1.36 SIG planning. I think this work would be well accepted by the SIG, but let's make sure.

srikalyan commented:
> thanks for the updates. The next step is to bring this up on the larger sig-node and in the 1.36 SIG planning. I think this work would be well accepted by the SIG, but let's make sure.

How do you recommend, I approach this?

srikalyan committed:
KEP-5759: Add reviewers and approvers to kep.yaml

- Add ffromani, derekwaynecarr, mrunalp as reviewers
- Add dchen1107 as approver (sig-node OWNERS)

srikalyan committed:
KEP-5759: Add PRR approval file and update approvers

- Add haircommander (Peter Hunt) as KEP approver
- Add PRR approval file for alpha stage with johnbelamaric as approver

johnbelamaric reviewed (commented):
One small comment otherwise PRR looks good.

@ffromani I guess this is something that will not be an issue if we use DRA for huge pages, assuming we say create each NUMA node as a device with a consumable capacity? We still need to fix this of course.

wendy-ha18 commented on keps/sig-node/5759-memory-manager-hugepages-verification/kep.yaml:
```diff
@@ -0,0 +1,46 @@
+  - "@srikalyan"
+owning-sig: sig-node
+participating-sigs: []
+status: provisional
```
```suggestion
status: implementable
```
Based on [Enhancement Freeze](https://github.com/kubernetes/sig-release/blob/master/releases/release_phases.md#enhancements-freeze) requirements, the KEP status set to `implementable`.

johnbelamaric commented:
FYI, for PRR just awaiting SIG approval, I have one nit above but I consider it non-blocking. kep.yaml update does need to happen too though.

srikalyan committed:
Apply suggestion from @wendy-ha18

Co-authored-by: Wendy Ha <139814343+wendy-ha18@users.noreply.github.com>

srikalyan committed:
Address PRR and design review feedback

- Move metrics from Beta to Alpha graduation criteria per ffromani's
  request to have observability available at alpha stage
- Change "TBD during alpha phase" to "Will be done during alpha phase"
  per johnbelamaric's nit on the upgrade/rollback testing question
- Add Alternative 3: Standalone NUMA-aware hugepages admission handler
  with pros/cons analysis per ffromani's suggestion
- Expand Alternative 1 with NUMA tracking limitation: without
  cpuset.mems enforcement, NUMA node allocation is unknown until
  container runtime, making per-pod tracking infeasible at admission
- Reframe race condition caveat to emphasize kubelet/workload contract
  breach rather than just startup failure timing
- Relax milestone timeline: beta v1.38, stable v1.40
- Remove sysfs availability from risk table (sysfs is a kubelet precondition)
- Recommend Option A (direct sysfs reading) with rationale
- Remove feature gate as safety mechanism framing throughout
- Remove hardcoded error message format (not a public API)
- Remove specific log format and alerting recommendation sections
- Simplify Events section to describe behavior without locking format
- Move conformance tests from GA to Beta criteria
- Update GA to "feature always enabled (feature gate removed)"
- Reword Upgrade/Downgrade without feature gate dependency
- Update rollback answer to reflect always-enabled at GA
- Replace speculative discrepancy metric with alpha evaluation plan

srikalyan commented on keps/sig-node/5759-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,697 @@
+### Real-World Example
+
+From [issue #134395](https://github.com/kubernetes/kubernetes/issues/134395),
+on an m6id.32xlarge instance with 2 NUMA nodes:
```
Added Alternative 3, Please resolve if you are good.f

srikalyan commented on keps/sig-node/5759-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,697 @@
+  **Why this is still valuable**: Without verification, the failure window spans
+  from pod scheduling to container startup (seconds to minutes). With verification,
+  the window is reduced to milliseconds between sysfs read and container start.
+  The vast majority of failures are prevented.
```
Strengthen the motivation here.

ffromani commented on keps/sig-node/5759-memory-manager-hugepages-verification/README.md:
````diff
@@ -0,0 +1,739 @@
+**Error message format**:
+```
+insufficient hugepages-2Mi on NUMA node(s) [0]: requested 4Gi, available 2Gi
+```
````
Does this mean the message format becomes part of the public API? This is not something we should do (and yes we did in the past and never turned out good)

ffromani commented on keps/sig-node/5759-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,739 @@
+### Upgrade / Downgrade Strategy
+
+**Upgrade**: No special handling required. The feature is additive and controlled
+by a feature gate. Existing pods are unaffected.
```
feature gate is transient in nature and it is not a good long-term safety mechanism

ffromani commented on keps/sig-node/5759-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,739 @@
+by a feature gate. Existing pods are unaffected.
+
+**Downgrade**: Disabling the feature gate returns to previous behavior where
+OS hugepage availability is not verified. No data migration needed.
```
ditto

srikalyan commented on keps/sig-node/5759-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,697 @@
+### Notes/Constraints/Caveats
+
+- **Race condition window**: A window exists between verification and actual
+  container startup where hugepages could be consumed by another process. This is
```
Fair point -- external processes stealing resources already violates kubelet's operating assumptions, so it's not a strong justification on its own. The primary motivation remains the Burstable pod tracking gap, which is a legitimate Kubernetes workload scenario. The caveats section has been updated to focus on the kubelet/workload contract breach framing.

srikalyan committed:
KEP-5759: Retarget alpha milestone to v1.37

Shift all milestones by one release cycle:
- alpha: v1.36 → v1.37
- beta: v1.38 → v1.39
- stable: v1.40 → v1.41

srikalyan committed:
KEP-5759: Clarify verification approach and trim implementation details

- Clarify dual-source verification: min(internal_free, os_free) per NUMA
  node to handle both untracked Burstable pod consumption and not-yet-faulted
  Guaranteed pod allocations
- Remove specific error message formats from KEP to avoid creating
  implicit API contracts
- Add user-observable behavior note pointing to event reason and metrics
  as the stable interface for identifying verification failures

srikalyan commented on keps/sig-node/5759-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,739 @@
+- Longer timeline
+
+The choice between options should be made during KEP review based on
+maintainability preferences and timeline considerations.
```
Already addressed in the current README at line 309 with Recommendation: Option A (Direct sysfs reading)
   and rationale

srikalyan commented on keps/sig-node/5759-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,739 @@
+|--------|------|-------------|
+| `memory_manager_hugepages_verification_total` | Counter | Total verification checks performed. Labels: `result` (success/failure), `hugepage_size` |
+| `memory_manager_hugepages_verification_failures_total` | Counter | Pods rejected due to insufficient OS-reported hugepages. Labels: `hugepage_size`, `numa_node` |
+| `memory_manager_hugepages_verification_latency_seconds` | Histogram | Time spent performing verification (buckets: 1ms to 100ms) |
```
Acknowledged — happy to defer to PRR reviewers on whether the latency histogram is needed at alpha. The verification is expected to be < 1ms so it may not provide actionable signal.

shiya0705 commented on keps/sig-node/5759-memory-manager-hugepages-verification/README.md:
```diff
@@ -0,0 +1,720 @@
+when hugepages aren't actually available, causing runtime failures.
+
+The enhancement adds verification by reading free hugepages from sysfs
+(`/sys/devices/system/node/node<N>/hugepages/hugepages-<size>kB/free_hugepages`)
```
Yes, it is more clear.

srikalyan commented:
@ffromani need your help to get this landed in next release at least. Can you please give another shot at the review please.
